### ISSUE TITLE
What have you done to this amazing package! 😔

### ISSUE THREAD

**Issue by spooksie** [ConsoleTVs/Charts#663]
First off, your package was amazing. It worked with Laravel really really well and was well documented. 

Whatever you did with the charisan update, totally ruined the repo. The documentation isn't as clear as it once and when you finish the installation guide. You are presented to go and learn charisan... Which that website has zero documentation or examples.

Is there any way to roll-back your repo, as it worked great. Now it's a cluster-fuck to get working. 

P.S All the guides are pointing to your old documentation and links are broken. Which is confusing to work with.

**Comment by ConsoleTVs** [comment]
On top of that, the old implementation was sync, meaning that if you had tons of data to load (like statistics) it would simply just... wait and load it.

Now you're free to load the page on time and fetch the data in the background... by default...

**Comment by peterchaula** [comment]
@spooksie why don't you go and improve the docs if you're not happy with them?

**Comment by vkyeswa** [comment]
@peterchaula how can @spooksie improve the documents yet not the one that made the changes?? Honestly I too get @spooksie's frustration, before (v6) the documentation was clear on how to use this package in laravel 

For example you could easily access the Chart in a controller like so,
`$chart = new SampleChart();
        $chart->labels(['First', 'Second', 'Third']);
        $chart->dataset('Sample', 'line', [1, 2, 3]);`

and displayed with a simple 
`view
{!! $chart->container() !!}
scripts
{!! $chart->script() !!}
`

But now it seems with v7 (which I get is a total rewrite as mentioned) all chart logic is mean't to be handled in the handler function
`public function handler(Request $request): Chartisan
    {
        return Chartisan::build()
            ->labels(['First', 'Second', 'Third'])
            ->dataset('Sample', [1, 2, 3])
            ->dataset('Sample 2', [3, 2, 1]);
    }`

and displayed with
`
    <script src="https://unpkg.com/echarts/dist/echarts.min.js"></script>
    
    <script src="https://unpkg.com/@chartisan/echarts/dist/chartisan_echarts.js"></script>
    
    <script>
      const chart = new Chartisan({
        el: '#chart',
        url: "@chart('sample_chart')",
      });
    </script>`

With this kind of change at least there should have been a comparison sort of documentation to show the huge difference between the two versions indicating breaking changes OR @ConsoleTVs would have created a new package **consoletv/chartisan** that would take over from this one coz honestly when one updates a package one expects bug fixes, new features but still inline with the previous logic and not to start changing all the workings of the package in all project controllers where it's being used.

That's my honest opinion, new version is great in terms of performance but the already existing users like @spooksie should get a **Breaking Change** warning before upgrading to v7

**Comment by peterchaula** [comment]
> how can @spooksie improve the documents yet not the one that made the changes

@vkyeswa I agree with you. The v7 release is lacking documentation and I had to dig through the code to get it to work. However, this is free and open software. We get it for free and we don't pay the project owner for his time. If you're not happy with the docs, why not go through the code, get a grasp of it, add some examples and create a PR?

**Comment by ConsoleTVs** [comment]
> should get a Breaking Change warning before upgrading to v7

Composer uses semantic versioning. If you see a MAJOR release, there are likely breaking changes in one way or another. I am not going to spend my free time explaining if that is the case or not since I expect developers to know such a basic thing when using 3rd party libraries.

On top of that, I do belive you do not understand why that handle method exists and why you cant do so it in a controller, that does not always correspond to a get method and it is purely sync. I do advise you to use chartisan/php since it does allow you to manually build the chart yourself in pure php and let you, as a dev, handle the routing yourself. Perhaps you prefer that.

**Comment by vkyeswa** [comment]
Well fellas at least we have all agreed on the issue at hand, like @peterchaula has said one has to do some digging to understand the changes made in v7. When issues are opened we are not complaining but only offering feedback so please @ConsoleTVs don't take issues raised like they are "attacks", nope just other devs having something to say about your work and it's good that you already believe that "we" don't understand why that handle method exists, so please help us understand .i.e through examples like you had in v6 docs.

By now all php developers know about the Semantic versioning of composer, and yet **read.me** files still existing coz documentation is powerful for any software acquisition. So if you get some free time, please update the documentation if not possible, don't mind, you've already done much to produce and maintain such an essential package it will be left to us who can to dig through and make sense of it.

_we all use/create packages to save on setup time_

**Comment by ConsoleTVs** [comment]
@vkyeswa Can you point out what you think needs more details in our documentation?

https://chartisan.dev/documentation

**Comment by vkyeswa** [comment]
I think the section [https://charts.erik.cat/guide/create_charts.html#create-the-chartisan-instance](url) needs some more detail like it had here in v6 [https://v6.charts.erik.cat/create_charts.html#create-a-chart-class](url). To be exact if the _create-the-chartisan-instance_ section could include _Database (Eloquent) examples_ like before to guide on how we can use this in our respective controllers.

**Comment by ConsoleTVs** [comment]
@vkyeswa I don't understand what you're seeking here. The documentation of that page clearly states: 

`You can know more about Chartisan at the Documentation page.` And a link to the documentation that I posted above.

Chartisan is responsible for the chart creation, this package ONLY handles some blade stuff, some routes, middlewares and a thin layer of abstraction on top of it.

So if you wan't to know how to create the data of the chart, visit chartisan's documentation.

On top of that, regarding your eloquent concerns, the Chartisan::build() just have 2 simple methods for you to use: `labels` and `dataset` where they both accept plain, vanilla, PHP. If you use eloquent, eloquent collections or collection instances provide a toArray() method you can use to convert it. 

```php
$users = Users::all()->countBy('name');

Chartisan::build()
    ->labels($users->keys()->toArray())
    ->dataset('Users with same name', $users->values()->toArray())
```

What's causing you confusion here? I can't really understand why a PHP array is causing that much trouble?

**Comment by vkyeswa** [comment]
@ConsoleTVs  I finally got the flow to the documentation, just mentioning the parts that were first confusing.
 If such a snippet 

>$ users = Users::all()->countBy('name');
> 
> Chartisan::build()
>     ->labels($users->keys()->toArray())
>     ->dataset('Users with same name', $users->values()->toArray())
> 

could be presented before the link of Chartisan documentation just for more customisation of the carts and further understanding how Chartisan works

> You can know more about Chartisan at the Documentation page.

coz it seems like the Chartisan documentation is in a default backend preference of NodeJS which kinda throws you off following from a PHP documentation before identifying that there is an option of changing preference... anyway that was my hurdle that kept me moving in circles

**Comment by arborrow** [comment]
As others have mentioned, for those of us using this in Laravel the switch to moving the configuration of things to JavaScript requires a fair amount of refactoring. I can understand the need and benefits of the new approach even if having more options or having it be more flexible has been a little confusing for me. 

As an example, I had a [dynamic number of datasets](https://github.com/arborrow/montserrat/blob/master/app/Http/Controllers/DashboardController.php#90) and a set of predefined set of colors that I would rotate through in a foreach loop for each dataset. It was great to be able to set the colors in PHP as I was creating the dataset. Now the chart api separates the data from the configuration.  As a result, I am trying to figure out what is the best way to get my custom colors into the JavaScript of the blade. I can have the controller pass them to the blade, but now I have to engage with JavaScript, try to determine how many datasets I have and loop through and assign the colors. How might an upgrade guide address such a scenario? 

Similarly with the same agc method, I was doing some queries to pull together data to create two charts - one for the number of donors and the other chart for the amount of the donations. Now, it is required to have separate charts for each of those (which isn't necessarily a bad thing); however, that means that in both places I need to do repetitive queries.

I did figure out how to protect the api endpoints with auth and AuthorizesRequests and that may been something that most Laravel developers would have figured out easily but I am not a full time developer so I can be a little slow. But essentially, I wanted the chart endpoint to behave as it had been behaving in the controller but getting $this->authorize('{name-of-role}') to work was another hurdle but ultimately not difficult to overcome.

As someone who also enjoys participating in open source, I am grateful for your time, efforts and insights.

**Comment by ConsoleTVs** [comment]
@arborrow I agree. I have a new version where you can use 90% of the JS api in plain PHP. It has been the result of merging both strategies, both allowing full JS control and "almost" full PHP control (limitations rely on serialization of closures - not possible). So yes, it is a bit of a mixed feeling and I get all the frustration. I must get back to work on this new version's documentation. I apologise

**Comment by arborrow** [comment]
No worries, let me know if there is anything I can do to be supportive. I'm happy to review or help with the new documentation. Hopefully the link to my v6 version gives a concrete example of what folks were running into. Similarly, I'd be happy to help with an upgrade guide.
